### PR TITLE
Ver 2.0.0

### PR DESCRIPTION
**Pull request by iuravic**
# Newspack Content Diff Migrator V.2

## Overview

The V.2 extends the Content Diff Migrator from a single-source migration tool into a multi-source system -- it is now able to migrate content from multiple remote WordPress sites independently, with newly added features and a significantly improved architecture and test coverage.

## New features

- Multiple source site migration support (with `--source-hostname` argument)
- [Newspack Migration Data Consistency Standard](https://newspackp2.wordpress.com/2025/09/24/migration-data-consistency-standard/) (MDCS) implementation
- Added default migration support for:
    - `wp_block` (synced patterns)
    - Simple Local Avatars's avatars
- Categories are no longer fully imported before migrating any posts, all of them are now created only on-the-fly which avoids importing any categories that are not used by the migrated content
- Major architectural refactoring for maintainability
- Completely rewritten and thoroughly more comprehensive test coverage (60+ test files, 28k+ lines of test code)
    - Unit tests for all the logic classes
    - Integration and E2E tests for all the core functionality
    - A CI job now runs tests on every push
- Refactored Run-state data (i.e. info about migration progress, for resume capability) and now clearly separated from logs (into its own `run-state/` sub-folder as formatted JSON/JSONL data files)
- CSV Reports are now automatically created at the end of every migration, which can be shared directly with Team and Publishers
- Consolidated logging, now using a single debug/action log file (one log file per command run), and switched to using Newspack Migration Tools' Monolog implementation
- Collation checking and fixing utils are now quite a lot faster and even more automated
- A new interactive CLI `index` command -- one command to rule them all -- for a convenient interactive menu of all available commands, descriptions and input of the arguments (required and optional)

## 📊 Line Count Breakdown: Where did these 28,541 Lines go?

This PR shows **+28,541 lines added** and **−7,795 lines were removed**... 😱 But don't be scared, it's really not that bad! 😅 

Here's where all those lines actually went!

| # Lines of Code | % of all Code | Functionality |
|--------|--------|--------|
| ~20,774 lines | **68%** | 🤖  Fully comprehensive test coverage with `599` tests (`369` unit + `230` integration/functional/E2E) covering the largest majority of migration scenarios and edge cases |
| ~4,800 lines | **17%** | ♻️ Refactored/reorganized existing code |
| ~3,000 lines | **11%** | ✅ New features and functionality |
| ~1,000 lines | **4%** | 📝 Documentation and configuration |

**The takeaways**
- ⭐  The PR prioritized **testing** and **code quality** over bloat! Most "new" lines are actually **tests (68% of all code)** plus **reorganized code (17%)**
- 🆕  The PR _**only added ~3k lines**_ of actual new code, for all the new functionality...
- 🗑️  ...while **removing ~8k lines** of legacy code.

This results in a **net improvement** in maintainability despite the large diff size 🙇 

### 🤖  Tests: ~20,774 Lines (68% of all additions!)

The comprehensive test coverage was greatly improved and is one of the biggest achievements:

**Unit Tests**: `8,727` lines, `369` unit tests

**Integration Tests**: `10,663` lines, `230` integration tests
- 18 comprehensive integration test classes, groupped by focus area
- E2E tests for all core functionality
- Includes `IntegrationTestCase.php` base class (402 lines)
- Notable: `CmdMigrateLiveContentMigrationDataConsistencyStandardTest.php` alone is 2,411 lines, to confirm the entire MDCS specification is working as expected.

**Test Fixtures**: 614 lines

### 📊 Code Coverage

A high test coverage achieved for the most important logic:

| Logic Class | Test Coverage |
|--------|--------|
| **BlockUpdater.php** | **95.93%** |
| **ContentDiffLogic.php** | **97.37%** |
| **DataImporter.php** | **92.97%** |
| **RunState.php** | **94.77%** | 

And this is the coverage for the most important command methods:

| Command Method | Test Coverage |
|--------|--------|
| **cmd_search_new_content_on_live** | 79.86% |
| **cmd_migrate_live_content** | 85.92% |
| **cmd_attribute_all_unattributed** | 92.86% |
| **cmd_attribute_match_local_to_live_tables** | 90.50% |
| **cmd_attribute_ids** | 79.61% |

### ♻️ Refactored Code: ~4,800 Lines (17%)

Code that was **moved/reorganized**, not newly written:

- **ContentDiffLogic** reduced from ~4,800 to 2,147 lines
  - Block updating logic → `BlockUpdater` (1,081 lines)
  - Data insertion logic → `DataImporter` (1,362 lines)
  - State management → `RunState` (543 lines)
  - Collation handling → `DB` (308 lines)
  
- **Command class** refactored for testability (2,228 lines, much reorganized)

**Deleted**: `7,795` lines removed
- Legacy test file: `3,842` lines
- Old Utils/PHP class: `40` lines
- Duplicate code, old patterns, and refactored methods: `~3,913` lines

### ✨ New Code: ~3,000 Lines (11%)

Actual **new functionality** written from scratch:

- `ContentDiffMigratorIndex` (interactive CLI): 374 lines
- `ReportCreator` (CSV generation): 440 lines
- `Logger` (unified logging): 211 lines
- `Progress` (progress meter): 102 lines
- `SLAHelper` (avatar support): 97 lines
- Multi-source collision handling: ~300 lines
- MDCS (Newspack [Migration Data Consistency Standard](https://newspackp2.wordpress.com/2025/09/24/migration-data-consistency-standard/)) implementation: ~400 lines
- Attribution commands rewrite: ~500 lines
- CSV report generation: ~300 lines
- Various smaller features: ~300 lines

### 📝 Documentation & Config: ~1,000 Lines (4%)

- README complete rewrite: ~500 lines
- PHPDoc comments throughout
- Config files (phpunit.xml.dist, phpcs.xml)

### 1. Mandatory `--source-hostname` Argument

Migration now supports importing content from multiple source sites simultaneously, contents of each site are tracked separately, each with its own source hostname.

This is a mandatory argument for the `search` and `migrate` commands, takes the source hostname of the site being migrated. There is also a safety check which will warn if a similar hostname has already been migrated (e.g. `example.com` vs `www.example.com`) and will prompt the user for confirmation to avoid accidental duplication.

**E.g. usage**: `--source-hostname=www.example.com`

### 2. "Old ID Meta" Key Changes

The new "old ID meta" is changed to "old ID **and** hostname meta" and now also contains the source hostname:
- before the meta was `'newspackcontentdiff_live_id' => 123`,
- after, the meta is `'newspackcontentdiff_oldid_www.example.com' => 123`

### 3. Unified `--data-dir` Argument

- **Before**: Mixed usage of `--export-dir` and `--import-dir`
- **After**: Always `--data-dir` for consistency
- **Benefit**: Simpler, clearer command syntax

And so, thanks to these naming updates, the required CLI arguments' names are now *identical* for the first search command, and for the second migrate command.

### 4. Command Naming Updates

Updated all commands to use the new repository naming convention:

**E.g. old command**: `wp newspack-content-migrator content-diff-search-new-content-on-live`
**New command**: `wp newspack-content-diff-migrator search-new-content-on-live`

### Multi-Source Migration Support

Now supports importing content from multiple WordPress sites simultaneously, each tracked with its own source hostname. Content from different sources is kept independent, with proper attribution and collision handling.

**Key capabilities**
- Track content from unlimited source hostnames
- Each source's old IDs stored in source-specific meta keys
- Intelligent merging for users and terms with same identifiers (unique columns/keys in the DB)
- Per-source attribution and reporting

### Migration Data Consistency Standard (MDCS)

Implements a comprehensive field-by-field specification for how content is detected and updated across migration runs. See [internal P2](https://newspackp2.wordpress.com/2025/09/24/migration-data-consistency-standard/) for full specification.

Also this behavior is documented in the README, all field updates and scenarios described.

### Code Organization & Separation of Concerns

**Before**: Single massive "Logic God Class" (~4800 lines) handling everything
**After**: Clean separation into simple(r), focused classes with single responsibilities

**Benefits**:
- Single Responsibility Principle, each class has one clear purpose
- All logic is now testable in isolation, plus integration/E2E coverage is enabled
- Dependency Injection, test-environment friendly constructors
- Reduced coupling

#### 1. `BlockUpdater`
Extracted all Gutenberg block ID update logic from `ContentDiffLogic` class.

#### 2. `DataImporter`

Refactored all post/user/term insertion logic from Logic and Command classes.

**Responsibilities**:
- Importing users
- Importing taxonomy trees
- Inserting posts
- Importing comments
- Importing terms
- Parent ID updates

#### 3. `RunState`

Manages the "migration state data" -- IDs which should be migrated, IDs which have been migrated -- used for progress tracking, and resumability in case of command interruption.

The migration state data is now stored separately from logs, in the `{--data-dir}/run-state/` sub-directory.

**Before**: Migration state data was scattered across Command and Logic classes, and mixed with logs
**After**: One simple dedicated class with a clean read/write API (`read_jsonl()`, `write_json()`, `append_jsonl()`), all data stored in a dedicated sub-directory

#### 4. `DB`

Extracted collation checking and fixing logic from `ContentDiffLogic` class, and moved it to a separate `DB` class.

**Improvements**:
- Now much faster, typically takes two minutes to run with new configuration
- Auto-cleanup of backup tables after collation fixes

#### 5. `Logger`
Unified logging interface wrapping Newspack Migration Tools' Monolog implementation, with handy customization.

**Features**:
- Dual output: CLI (with summary) + File (with full context)
- Singleton pattern for logging intialization and sharing across services/classes, test-environment friendly

#### 6. `Progress`
Super small utility for progress meter output during long operations (percentage-based).

#### 7. `ReportCreator`
Generates CSV reports from JSONL run-state data.

#### 8. `SLAHelper`
Super small utility for setting Simple Local Avatars's avatars. It is a standalone class, not requiring SLA plugin as dependency, for speed and simplicity.

#### 9. `ContentDiffMigratorIndex`

The new `index` command is just a fun add-on, and an optional extra which lets you run *any* command in interactive mode. Instead of memorizing command names and arguments, you may wish to run it and:
- list all commands, with their descriptions
- chose which command(s) to run
- input all parameters interactively, required and optional

image

But, once again, this is just a convenient optional extra, which is also why it's in a command class of its own.

#### `ContentDiffLogic` (Reduced from ~4800 to 1,500 lines)
Focused on core content diff logic only:
- Post field comparison
- ID filtering and matching
- Block ID coordination
- Term operations

**Removed** (moved to appropriate classes):
- All block updating → `BlockUpdater`
- All data insertion → `DataImporter`
- All state management → `RunState`
- All collation handling → `DB`

#### `ContentDiffMigrator`
Command class cleaned and made integration-testable:
- Removed business logic (moved to Logic classes)
- Made test-environment friendly for comprensive integration tests

### Attribution Commands

When you run a CDiff migration, the imported content gets the "old ID and hostname meta" assigned to it. The following WP data objects can have this meta assigned to them, stored in the corresponding WP metas tables:
- Posts/Pages/Attachmewnts/CPTs (stored in `wp_postmeta`)
- Users (stored in `wp_usermeta`)
- Terms (stored in `wp_termmeta`)

But what happens if you clone a site, or have performed some independent custom migration with the site's content, and you need to run a content diff _**including**_ that content?

There are three easy and conveniently different ways (sub-commands) how you can "attribute" (assign) the "old ID and source hostname" metas to it.

And just to clarify, to **"attribute some content to a source hostname"** simply means to assign "old ID and hostname metas" to that content, nothing more -- so that the CDiff plugin can track that content to the original source hostname, properly compare it to live tables, and do content refreshes. These metas are what makes the multi-source work.

These three commands all do the very same thing (assign metas), but in slightly different ways, for your convenience.

1. `attribute-all-unattributed` command
    - will attribute **ALL** unattributed local content to a source hostname
    - **Use case**: perfect to run **immediately** after cloning a site and it will assign the metas to all the existing content, and it doesn't need the live tables
2. `attribute-match-local-to-live-tables` command
    - compares the unattributed local content with live tables, and automatically finds matches and sets the "old ID and hostname" meta
    - **Use case**: You cloned the live DB to your local site, but also some custom content was created there too (e.g. Newspackification stuff). And you need to attribute just the original cloned content to the source hostname, not all of it
3. `attribute-ids` command
    - takes specific IDs of `posts` (and all CPTs), `users`, and/or `terms` and attributes just those objects to a source hostname
    - **Use case**: Some different custom migration was done in parallel for some reason, and then you wish to also run CDiff on that content. So you first assign the custom-migrated IDs metas, so that CDiff knows how to do the content refresh. This is very much an edge case, made for "just in case".

All of these commands will also generate simple **attribution reports** -- timestamped CSVs in `{--data-dir}/reports/`:
- `attributed_posts_{timestamp}.csv` - Posts and attachments which were attributed
- `attributed_users_{timestamp}.csv` - Users attributed
- `attributed_terms_{timestamp}.csv` - Terms attributed

### CSV Reports & Migration Summary

Every single migration run now automatically produces formatted CSVs with all migrated content, ready to be shared with the team or customer for QA and analysis.

**Generated reports** (in `{--data-dir}/reports/`):
- `posts.csv` - All migrated/reimported posts, pages, attachments, CPTs
- `users.csv` - All imported/merged/modified users
- `terms.csv` - All imported/merged terms (categories, tags, custom taxonomies)

### Improved logging
- Monolog-based structured logging (via Newspack Migration Tools)
- Single consolidated debug/action log file per command, instead of previously multiple log files per command
- Convenient Dual logging: CLI gets a concise brief output (which doesn't overload the screen), while the action/debug file log always gets fully verbose and standardized messages with all the context data

### Term ID Mapping Meta

Terms now receive `newspackcontentdiff_oldid_{hostname}` termmeta during import, matching the pattern used for posts and users.

**Benefits**:
- Reliable term identification across migrations
- Consistent meta key naming across all object types
- Enables efficient bulk lookups

### Simple Local Avatars Support

Added support for Simple Local Avatars plugin:
- Avatar attachments migrated with users
- Avatar meta updated on consecutive runs
- Follows [MDCS](https://newspackp2.wordpress.com/2025/09/24/migration-data-consistency-standard/) field update rules

### Modified Post Reimport

When modified posts are marked modified for a full reimport, the existing `wp_posts.ID` is preserved, and the post gets reimported with that same ID.

### Featured Image Handling

Some edge cases were added and covered, which were missing from v.1, and discovered thanks to comprehensive new test suite.

### Removed Block for Legacy Guest Authors

Incorporates [PR#8](https://github.com/Automattic/newspack-content-diff-migrator/pull/8) and removes the previous hard-coded block for CoAuthors Plus's Guest Authors.

Though legacy Guest Authors are not migrated by default, they may now be optionally migrated using the custom `--post-types-csv` argument (a memo on how to do that is added to README).

### Multi-Source Collision Handling

Data rows with same unique keys/columns are gracefully "merged", i.e. reused. For example, if there is already a User or a Term record with identical unique columns, the migrator will reuse ("merge") those rows in the migration, and it will display a full warning in the debug/action log to let you know that this has been done.

Additionally, if a user was "merged", the reports CSV contains a `status` column with possible values (freshly) "imported", or "merged" -- see README for full specs on CSVs.

### Detailed Error Logging

Stack traces and context in debug/action log, with full details.

At the same time CLI screen logging messages remain minimal with clear warnings and pointing to the debug log for full contenxt, but without unnecessary outputting of context to crowd the screen.

### Complete Test Rewrite

_**Before**_ -- 1 legacy test file/class with 3,842 lines where some tests were becoming very hard to maintain and extend, which exposed some anti-patterns in the code -- mainly **tight coupling** and methods and classes which wanted to do **too many things**.

_**After**_ -- **Unit tests** for all the logic classes' code, and **Integration/E2E tests** for all the core migration functionality, clearly grouped and much, much more maintainable.

### Unit tests

All key logic classes are thoroughly unit tested, with comprehensive coverage of all their functionality.

**Total unit tests:** `369` unit tests, `8,727` lines of code.

| Class | No. Unit Tests | Lines of Test Code |
|--------|--------|--------|
| `tests/unit/Logic/BlockUpdaterTest.php` | `71` unit tests | `1,870` lines of code |
| `tests/unit/Logic/ContentDiffLogicTest.php` | `140` unit tests | `2,429` lines of code |
| `tests/unit/Logic/DataImporterTest.php` | `89` unit tests | `2,749` lines of code |
| `tests/unit/Logic/RunStateTest.php` | `46` unit tests | `903` lines of code |
| `tests/unit/Utils/ReportCreatorTest.php` | `23` unit tests | `776` lines of code |

### Integration Tests (18 test classes, `230` integration tests, ~11k lines)

Integration tests are grouped into multiple test classes, by the functionality/topic, to cover all aspects of migrations. It tests the plugin in a real-world scenario, and utilizes E2E (end-to-end) testing.

**Integration Test Groups/Classes**:

| What the Group Tests | No. Integration Tests | Lines of Code | Class/Group |
|--------|--------|--------|--------|
| Attachment migration | `7` tests | `253` | `CmdMigrateLiveContentAttachmentsTest` |
| Attribution commands | `16` tests | `663` | `CmdAttributeTest` |
| Block ID updates | `12` tests | `537` | `CmdMigrateLiveContentBlocksTest` |
| Comment migration | `7` tests | `436` | `CmdMigrateLiveContentCommentsTest` |
| Error recovery | `10` tests | `290` | `CmdMigrateLiveContentErrorHandlingTest` |
| Featured image handling | `4` tests | `153` | `CmdMigrateLiveContentFeaturedImageTest` |
| Full E2E migration flows | `4` tests | `204` | `CmdMigrateLiveContentE2ETest` |
| Listing hostnames | `8` tests | `180` | `CmdListSourceHostnamesTest` |
| Newspack Migration Data Consistency Standard implementation | `47` tests |`2,412` | `CmdMigrateLiveContentMigrationDataConsistencyStandardTest` |
| Modified post reimport | `6` tests | `444` | `CmdMigrateLiveContentPostsModifiedTest` |
| Multi-run and repeated runs scenarios | `6` tests | `252` | `CmdMigrateLiveContentRerunAndIdempotencyTest` |
| Multi-source merging | `9` tests | `813` | `CmdMigrateLiveContentMultiSourceCollisionTest` |
| Post migration | `12` tests | `502` | `CmdMigrateLiveContentPostsTest` |
| Search command | `13` tests | `432` | `CmdSearchNewContentOnLiveTest` |
| Source tracking | `4` tests | `133` | `CmdMigrateLiveContentSourceHostnameTest` |
| State tracking | `21` tests | `893` | `CmdMigrateLiveContentRunStateTest` |
| Taxonomy migration | `34` tests |`1,410` | `CmdMigrateLiveContentTaxonomyTest` |
| User migration | `10` tests | `438` | `CmdMigrateLiveContentUsersTest` |

### Test Fixtures

- **Block fixtures** - Gutenberg blocks for ID update testing, `10` HTML files, `67` lines of code
- **Live table fixtures** - DB data for integration tests, `5` JSON files, `547` lines of code

## Technical Improvements

1. **Full CI test suite** now runs on every push to repo
2. **PHP 8 Strict Typing**: Type hints throughout
3. **Excluded Deprecated WordPress Sniffs**: For JS/CSS files
4. **📋 README Updates**: Complete rewrite with better organization, workflow examples, full [MDCS](https://newspackp2.wordpress.com/2025/09/24/migration-data-consistency-standard/) specification included, auto-generated reports docs

---

## Next Steps After Merge

- [ ] Create release tag
- [ ] Update Knife integration

### CODE REVIEW

File: src/Logic/DB.php [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+			}
+
+			if ( $sleep_between_batches > 0 ) {
+				sleep( $sleep_between_batches );
+			}
+		}
+
+		// Delete backup table after successful copy.
+		// phpcs:ignore -- query fully sanitized.
+		$drop_result = $this->wpdb->query( "DROP TABLE IF EXISTS $backup_table" );
```
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Logic/DB.php]: Applied in [2bd4b35](https://github.com/Automattic/newspack-content-diff-migrator/pull/9/commits/2bd4b35942adb2bf88279591738817f05aba2592)

File: src/Utils/DB.php [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+			$insert_sql = "INSERT INTO `{$source_table}`({$table_columns}) SELECT {$table_columns} FROM {$backup_table} LIMIT {$limiter['start']}, {$limiter['limit']}";
+			// phpcs:ignore -- query fully sanitized.
+			$insert_result = $this->wpdb->query( $insert_sql );
+
+			if ( ( false !== $insert_result ) && ( 0 !== $insert_result ) ) {
+				$limiter['start'] = $limiter['start'] + $limiter['limit'];
+			} else {
+				$db_error = ( '' != $this->wpdb->last_error ) ? 'DB error message: ' . $this->wpdb->last_error : 'No DB error message available -- check error and debug logs.';
+				Logger::instance()->log( Logger::OUTPUT_BOTH, LogLevel::ERROR, sprintf( "Got up to (not including) %s. Failed running SQL '%s'. %s", $limiter['start'], $insert_sql, $db_error ) );
+			}
```
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Utils/DB.php:316]: Applied 👍

File: src/Utils/SLAHelper.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	 *
+	 * @return array Map of user_id => unserialized avatar meta array.
+	 */
+	public function get_avatar_meta_by_user_id( string $table_prefix ): array {
+		// phpcs:disable -- WordPress.DB.DirectDatabaseQuery.DirectQuery WordPress.DB.DirectDatabaseQuery.NoCaching WordPress.DB.PreparedSQL.InterpolatedNotPrepared.
+		$results = $this->wpdb->get_results(
+			$this->wpdb->prepare(
+				"SELECT user_id, meta_value FROM {$table_prefix}usermeta WHERE meta_key = %s",
+				self::AVATAR_META_KEY 
+			),
```
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Utils/SLAHelper.php:49]: Absolutely !!! Thank you for catching this 🙇 Applied in [eb31f9d](https://github.com/Automattic/newspack-content-diff-migrator/pull/9/commits/eb31f9d7c14d11c495acc3c55f63ba7a74c61308).

File: src/Utils/SLAHelper.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+		$meta_value = [
+			'media_id' => $attachment_id,
+			'full'     => $attachment_url,
+			'blog_id'  => get_current_blog_id(),
+		];
+
+		$updated = update_user_meta( $user_id, self::AVATAR_META_KEY, $meta_value );
+		if ( false === $updated ) {
+			return new WP_Error( 'update_failed', sprintf( 'Failed to update avatar for user ID %d.', $user_id ) );
```
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Utils/SLAHelper.php]: Applied in [a98347e](https://github.com/Automattic/newspack-content-diff-migrator/pull/9/commits/a98347efe706b558a59e4be92c4a3e930dc876df).

File: src/Utils/SLAHelper.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+	/**
+	 * Sets a Simple Local Avatar for a user.
+	 *
+	 * @param int $user_id       User ID.
+	 * @param int $attachment_id Attachment ID of the avatar image.
+	 *
+	 * @return bool|WP_Error True on success, WP_Error on failure.
+	 */
+	public function set_user_avatar( int $user_id, int $attachment_id ): bool|WP_Error {
```
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Utils/SLAHelper.php:69]: Applied in [a98347e](https://github.com/Automattic/newspack-content-diff-migrator/pull/9/commits/a98347efe706b558a59e4be92c4a3e930dc876df).

File: src/Utils/ReportCreator.php [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+			$post_type = 'post';
+			if ( isset( $posts_by_id_new[ $id_new ]['post_type'] ) ) {
+				$post_type = $posts_by_id_new[ $id_new ]['post_type'];
+			} else {
+				// Try to get post type from database.
+				$wp_post = get_post( $id_new );
+				if ( $wp_post ) {
+					$post_type = $wp_post->post_type;
+				}
+			}
```
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Utils/ReportCreator.php:129]: This is a nice one, though not crucial -- it's just the Reports class which creates CSVs at the end of the migration. But will still apply this 👍

File: src/Utils/SLAHelper.php [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	 *
+	 * @return array Map of user_id => unserialized avatar meta array.
+	 */
+	public function get_avatar_meta_by_user_id( string $table_prefix ): array {
+		// phpcs:disable -- WordPress.DB.DirectDatabaseQuery.DirectQuery WordPress.DB.DirectDatabaseQuery.NoCaching WordPress.DB.PreparedSQL.InterpolatedNotPrepared.
+		$table_prefix_escaped = esc_sql( $table_prefix );
+		$results              = $this->wpdb->get_results(
+			$this->wpdb->prepare(
+				"SELECT user_id, meta_value FROM {$table_prefix_escaped}usermeta WHERE meta_key = %s",
+				self::AVATAR_META_KEY
```
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Utils/SLAHelper.php]: Oh yes, great catch a 100% 👍  I'll also be double-checking **ALL** the raw queries made throughout this plugin for this.

File: src/Command/ContentDiffMigratorIndex.php [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+	public static function register_command(): void {
+		WP_CLI::add_command(
+			'newspack-content-diff-migrator index',
+			[ __CLASS__, 'cmd_index' ],
+			[
+				'shortdesc' => 'Interactive command selector with guided argument input.',
+				'longdesc'  => 'Displays an interactive menu to select and execute any newspack-content-diff-migrator command with step-by-step prompts for all required and optional arguments.',
+			]
+		);
+	}
```
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Command/ContentDiffMigratorIndex.php:32]: Will check this for future compatibility 👍

File: src/Logic/DB.php [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+				continue;
+			}
+			
+			$tables_checked++;
+			$live_tablename = $live_table_prefix . $table;
+			if ( ! in_array( $live_tablename, $all_tables ) ) {
+				throw new \RuntimeException( sprintf( 'Core WP DB table %s not found.', $live_tablename ) ); // phpcs:ignore WordPress.Security.EscapeOutput.ExceptionNotEscaped
+			}
+			$tables_validated++;
+		}
```
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Logic/DB.php]: Applied in [1605377](https://github.com/Automattic/newspack-content-diff-migrator/pull/9/commits/1605377c459118cb5881536aef8bb6f2134e3b1b).

File: src/Logic/DB.php [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+	 *
+	 * @throws \RuntimeException In case tables don't exist or collations don't match.
+	 */
+	public function validate_db_tables( string $live_table_prefix, array $skip_tables = [] ): void {
+		// Check whether core WP DB tables that exist in the installation are present in used DB.
+		$all_tables       = $this->get_all_db_tables();
+		$tables_checked   = 0;
+		$tables_validated = 0;
+		
+		foreach ( self::CORE_WP_TABLES as $table ) {
```
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Logic/DB.php]: Applied in [1605377](https://github.com/Automattic/newspack-content-diff-migrator/pull/9/commits/1605377c459118cb5881536aef8bb6f2134e3b1b).

File: src/Utils/ReportCreator.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Utils/ReportCreator.php:129]: It doesn't seem necessary for migration logic to store post_type in run-state at this point. Seeing how this is a post-migration utility, evaluating that it's OK for this query to remain used in ReportsCreator. However now optimized and removed usage of `get_post()` and replaced it with a slim direct query in [2399f0e](https://github.com/Automattic/newspack-content-diff-migrator/pull/9/commits/2399f0e64c1da8845b639a6e2dd3bf2dadcb5207).

File: src/Utils/SLAHelper.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Utils/SLAHelper.php]: Applied in [c1e579c](https://github.com/Automattic/newspack-content-diff-migrator/pull/9/commits/c1e579c3618f593ddd9f7610f91b33b3aec673cb), trough out the entire code.

File: src/Command/ContentDiffMigratorIndex.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Command/ContentDiffMigratorIndex.php:32]: Applied here [3669cd8](https://github.com/Automattic/newspack-content-diff-migrator/pull/9/commits/3669cd8c0dfa02ea5c9fbd939576898b91f6fb1b).

File: src/Command/ContentDiffMigratorIndex.php [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+
+	/**
+	 * Main interactive index command.
+	 *
+	 * @param array $pos_args   Positional CLI args (unused).
+	 * @param array $assoc_args Associative CLI args (unused).
+	 */
+	public function cmd_index( array $pos_args, array $assoc_args ): void { // phpcs:ignore Generic.CodeAnalysis.UnusedFunctionParameter
```
[the remark this answers is not in the document]
Comment (iuravic) [reply src/Command/ContentDiffMigratorIndex.php:40]: Applied here [3669cd8](https://github.com/Automattic/newspack-content-diff-migrator/pull/9/commits/3669cd8c0dfa02ea5c9fbd939576898b91f6fb1b).

File: src/Utils/DB.php [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+	 *
+	 * @param string $prefix           Live table prefix.
+	 * @param string $table            The Core WP Table to address.
+	 * @param int    $total_size_bytes Total size of all tables being fixed (for speed determination).
+	 *
+	 * @throws \RuntimeException Throws various exceptions if unable to complete required SQL operations.
+	 */
+	public function copy_table_data_using_proper_collation( string $prefix, string $table, int $total_size_bytes = 0 ): void {
+		// Auto-determine speed based on total size of tables being fixed.
+		$four_gb_in_bytes = 4 * 1024 * 1024 * 1024;
```
Comment (eddiesshop) [inline src/Utils/DB.php]: Can we switch this to a constant?

File: src/Utils/DB.php [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+		// phpcs:ignore -- query fully sanitized.
+		$drop_result = $this->wpdb->query( "DROP TABLE IF EXISTS $backup_table" );
+		if ( false === $drop_result ) {
+			Logger::instance()->log( Logger::OUTPUT_BOTH, LogLevel::WARNING, sprintf( "Failed to drop backup table '%s'.", $backup_table ) );
+		} else {
+			Logger::instance()->log( Logger::OUTPUT_BOTH, LogLevel::DEBUG, sprintf( "Deleted backup table '%s'.", $backup_table ) );
+		}
+
+		// Sleep after table is complete (for small datasets).
+		if ( $sleep_after_table > 0 ) {
```
Comment (eddiesshop) [inline src/Utils/DB.php]: This is never changed, it's always set to 10, so this condition will always be true. Did you mean to always have a sleep between iterations?

**Review by eddiesshop — COMMENTED** [review]
I'm including a diff with some suggestions for this class. Some are for clearer documentation, the other are for some PHPCS issues that appear for me.

[diff.patch](https://github.com/user-attachments/files/24944233/diff.patch)

File: src/Utils/DB.php (hunk 12 above)
Comment (iuravic) [reply src/Utils/DB.php]: Good one. Done in https://github.com/Automattic/newspack-content-diff-migrator/pull/9/commits/80452a98cb5dfdf135f5f0903a01bc282c51f643.

File: src/Utils/DB.php (hunk 13 above)
Comment (iuravic) [reply src/Utils/DB.php]: I agree, thanks, no need to have the if there. Now removed in https://github.com/Automattic/newspack-content-diff-migrator/pull/9/commits/c4e39c3d6b609f156c0c16485461b11b9fbceeba.

**Comment by iuravic** [comment]
Hi @eddiesshop 👋 I committed both suggestions for the DB util class, thanks for those. However I'm getting a not found for the linked "diff.patch" [here](https://github.com/Automattic/newspack-content-diff-migrator/pull/9/#pullrequestreview-3724077170). Serves me right when taking too long to answer to reviews! Was there some content lost there, or did the convo about the two points contain them already? Thanks.

**Comment by eddiesshop** [comment]
> Hi @eddiesshop 👋 I committed both suggestions for the DB util class, thanks for those. However I'm getting a not found for the linked "diff.patch" [here](https://github.com/Automattic/newspack-content-diff-migrator/pull/9/#pullrequestreview-3724077170). Serves me right when taking too long to answer to reviews! Was there some content lost there, or did the convo about the two points contain them already? Thanks.

Let me see if I can find it. But if not, no worries, it can be handled later on.
